### ISSUE TITLE
SMTP support for OIDC

### ISSUE THREAD

**Issue by chibenwa** [linagora/james-project#4446]
## Why? 

OAuth is an industry standard and IMAP OAuth support is a recuring question we get in the community.

This is specified in https://datatracker.ietf.org/doc/html/rfc7628 .

## How? 

We will expose an option (disabled by default) to enable OAuth authentication in SMTP.

The SMTP configuration should:
  - URL of the OIDC supplier
  - Name of the claim acting as the login (email address)
  - JWKS URL of the OIDC provider

Exemple:

```
    <smtpserver enabled="true">
        <jmxName>smtpserver-TLS</jmxName>
        <bind>0.0.0.0:465</bind>
        <connectionBacklog>200</connectionBacklog>
        <tls socketTLS="true" startTLS="false">
            <privateKey>file://conf/private.key</privateKey>
            <certificates>file://conf/certs.self-signed.csr</certificates>
        </tls>
        <!--- ... -->
        <smtpGreeting>Apache JAMES awesome SMTP Server</smtpGreeting>
        <authentication>
            <plainAuthEnabled>true|false</plainAuthEnabled> <!-- defaults to true -->
            <oauth> <!-- ommiting this block would disable oauth. All fields are compulsory -->
                <jwksURL>https://example.com/jwks</jwksURL> 
                <oidcConfigurationURL>https://example.com/.well-known/openid-config</oidcConfigurationURL>
                <claim>mailAddress</claim>
            </oauth>
        </authentication>
        <handlerchain>
            <handler class="org.apache.james.smtpserver.fastfail.ValidRcptHandler"/>
            <handler class="org.apache.james.smtpserver.CoreCmdHandlerLoader"/>
        </handlerchain>
    </smtpserver>
```

of course the authentication part of the configuration file should be backed by a configuration POJO .

Implement a OAUTH mechanism in `AuthCmdHandler`, using the right `JWTVerifier` as created from the configuration. https://github.com/auth0/jwks-rsa-java can be used to get easily the public key out of a JWK endpoint.

## Examples from the spec

TLDR: a first roundtrip is needed for the client to discover id provider location:

```
   This example shows an authorization failure in an SMTP exchange.  TLS
   negotiation is not shown, but as noted above, it is required for the
   use of bearer tokens.

[connection begins]
S: 220 mx.example.com ESMTP 12sm2095603fks.9
C: EHLO sender.example.com
S: 250-mx.example.com at your service,[172.31.135.47]
S: 250-SIZE 35651584
S: 250-8BITMIME
S: 250-AUTH LOGIN PLAIN OAUTHBEARER
S: 250-ENHANCEDSTATUSCODES
S: 250 PIPELINING
[Negotiate TLS...]
C: AUTH OAUTHBEARER bix1c2VyPXNvbWV1c2VyQGV4YW1wbGUuY29tLAFhdXRoPUJlYXJl
       ciB2RjlkZnQ0cW1UYzJOdmIzUmxja0JoZEhSaGRtbHpkR0V1WTI5dENnPT0BAQ==
S: 334 eyJzdGF0dXMiOiJpbnZhbGlkX3Rva2VuIiwic2NoZW1lcyI6ImJlYXJlciBtYWMiL
       CJzY29wZSI6Imh0dHBzOi8vbWFpbC5leGFtcGxlLmNvbS8ifQ==
C: AQ==
S: 535-5.7.1 Username and Password not accepted. Learn more at
S: 535 5.7.1 http://support.example.com/mail/oauth
[connection continues...]

   The initial client response is:

   n,user=someuser@example.com,^A
   auth=^A^A

   The decoded server error response is:

  {
  "status":"invalid_token",
  "scope":"example_scope",
  "openid-configuration":"https://example.com/.well-known/openid-config"
  }

   The server returned an error message in the 334 SASL message; the
   client responds with the required dummy response, and the server
   finalizes the negotiation.
```

Note that an empty token allows discovering the OIDC configuration.

Having discovered the location of the OpenId server the client can then get its token there and come back:

```
   The same credential used in an SMTP exchange is shown below.  Again,
   this example assumes that TLS is already established per the bearer
   token specification requirements.

   [connection begins]
   S: 220 mx.example.com ESMTP 12sm2095603fks.9
   C: EHLO sender.example.com
   S: 250-mx.example.com at your service,[172.31.135.47]
   S: 250-SIZE 35651584
   S: 250-8BITMIME
   S: 250-AUTH LOGIN PLAIN OAUTHBEARER
   S: 250-ENHANCEDSTATUSCODES
   S: 250-STARTTLS
   S: 250 PIPELINING
   [Negotiate TLS...]
   C: t1 AUTH OAUTHBEARER bixhPXVzZXJAZXhhbXBsZS5jb20sAWhvc3Q9c2Vy
         dmVyLmV4YW1wbGUuY29tAXBvcnQ9NTg3AWF1dGg9QmVhcmVyIHZGOWRmd
         DRxbVRjMk52YjNSbGNrQmhiSFJoZG1semRHRXVZMjl0Q2c9PQEB
   S: 235 Authentication successful.
   [connection continues...]

   The decoded initial client response is:

   n,a=user@example.com,^Ahost=server.example.com^Aport=587^A
   auth=Bearer vF9dft4qmTc2Nvb3RlckBhbHRhdmlzdGEuY29tCg==^A^A
```

## Definition of done

 - **MPT tests** that exercise oauth SASL mechanism

JWKS endpoint can be mocked using Mock HTTP server.

## Limitations

 - No revocation of token is supported so far (valid signed tokens can always be used)
 - Revoking a token to not end the SMTP sessions using that token.

**Comment by quantranhong1999** [comment]
What is MPT test? Is there any example in James?

**Comment by chibenwa** [comment]
MPT = Mail Protocol Tester.

We use it for (some) tests against IMAP, SMTP

Look within `/mpt/impl/imap/core/src/resources/script/...`
